bin/barclamp_lib.rb

```diff
@@ -140,12 +93,363 @@ def usage(rc)
+  r.body = data if data
+
+  res = h.request r
+
+  debug "(r) hostname: #{uri.host}:#{uri.port}"
+  debug "(r) request: #{uri.path}"
+  debug "(r) method: #{req::METHOD}"
+  debug "(r) return code: #{res.code}"
+  debug "(r) return body: #{res.body}"
+  res.each_header do |h, v|
+    debug "(r) return #{h}: #{v}"
+  end
+
+  if res["www-authenticate"]
+    digest = Net::HTTP::DigestAuth.new
+    auth = digest.auth_header uri, res["www-authenticate"], req::METHOD
+
+    r = req.new uri.request_uri, @headers
+    r.body = data if data
+    r.add_field "Authorization", auth
+
+    res = h.request r
+
+    debug "(a) hostname: #{uri.host}:#{uri.port}"
+    debug "(a) request: #{uri.path}"
+    debug "(a) method: #{req::METHOD}"
+    debug "(a) return code: #{res.code}"
+    debug "(a) return body: #{res.body}"
+    res.each_header do |h, v|
+      debug "(a) return #{h}: #{v}"
+    end
+  end
+
+  res
+end
+
+def get_json(path)
+  uri = URI.parse("http://#{@hostname}:#{@port}/crowbar/#{@barclamp}/1.0#{path}")
+  res = authenticate(Net::HTTP::Get,uri)
+
+  debug "(g) hostname: #{uri.host}:#{uri.port}"
+  debug "(g) request: #{uri.path}"
+  debug "(g) return code: #{res.code}"
+  debug "(g) return body: #{res.body}"
+
+  return [res.body, res.code.to_i] if res.code.to_i != 200
+
+  struct = JSON.parse(res.body)
+
+  debug "(g) JSON parse structure = #{struct.inspect}"
+
+  return [struct, 200]
+end
+
+def post_json(path, data)
+  uri = URI.parse("http://#{@hostname}:#{@port}/crowbar/#{@barclamp}/1.0#{path}")
+  res = authenticate(Net::HTTP::Post,uri,data)
+
+  debug "(post) hostname: #{uri.host}:#{uri.port}"
+  debug "(post) request: #{uri.path}"
+  debug "(post) data: #{@data}"
+  debug "(post) return code: #{res.code}"
+  debug "(post) return body: #{res.body}"
+
+  [res.body, res.code.to_i]
+end
+
+def put_json(path, data)
+  uri = URI.parse("http://#{@hostname}:#{@port}/crowbar/#{@barclamp}/1.0#{path}")
+  res = authenticate(Net::HTTP::Put,uri,data)
+
+  debug "(put) hostname: #{uri.host}:#{uri.port}"
+  debug "(put) request: #{uri.path}"
+  debug "(put) data: #{@data}"
+  debug "(put) return code: #{res.code}"
+  debug "(put) return body: #{res.body}"
+
+  [res.body, res.code.to_i]
+end
+
+def delete_json(path)
+  uri = URI.parse("http://#{@hostname}:#{@port}/crowbar/#{@barclamp}/1.0#{path}")
+  res = authenticate(Net::HTTP::Delete,uri)
+
+  debug "(d) hostname: #{uri.host}:#{uri.port}"
+  debug "(d) request: #{uri.path}"
+  debug "(d) return code: #{res.code}"
+  debug "(d) return body: #{res.body}"
+
+  [res.body, res.code.to_i]
+end
+
+def list
+  struct = get_json("/")
+
+  if struct[1] != 200
+    ["Failed to talk to service list: #{struct[1]}: #{struct[0]}", 1]
+  elsif struct[0].nil? or struct[0].empty?
+    ["No current configurations", 0]
+  else
+    out = ""
+    struct[0].sort.each do |name|
+      out = out + "\n" if out != ""
+      out = out + "#{name}"
+    end
+    [out, 0]
+  end
+end
+
+def api_help
+  struct=get_json("/help")
```

Comment:
Style/SpaceAroundOperators: Surrounding space missing for operator `=`. (https://github.com/bbatsov/ruby-style-guide#spaces-operators)